backend/httpclient/datasource_metrics_middleware.go

```diff
@@ -0,0 +1,153 @@
+package httpclient
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const kb = 1024
+const mb = kb * kb
+const gb = mb * kb
+
+var (
+	datasourceRequestCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "plugins",
+			Name:      "datasource_request_total",
+			Help:      "A counter for outgoing requests for an external data source",
+		},
+		[]string{"datasource", "datasource_type", "code", "method", "secure_socks_ds_proxy_enabled"},
+	)
+
+	datasourceRequestHistogram = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "plugins",
+			Name:      "datasource_request_duration_seconds",
+			Help:      "histogram of durations of outgoing external data source requests sent from Grafana",
+			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 25, 50, 100},
+		}, []string{"datasource", "datasource_type", "code", "method", "secure_socks_ds_proxy_enabled"},
+	)
+
+	datasourceResponseHistogram = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "plugins",
+			Name:      "datasource_response_size_bytes",
+			Help:      "histogram of external data source response sizes returned to Grafana",
+			Buckets: []float64{128, 256, 512, 1 * kb, 2 * kb, 4 * kb, 8 * kb, 16 * kb, 32 * kb, 64 * kb, 128 * kb, 256 * kb, 512 * kb, 1 * mb,
+				2 * mb, 4 * mb, 8 * mb, 16 * mb, 32 * mb, 64 * mb, 128 * mb, 256 * mb, 512 * mb, 1 * gb,
+				2 * gb, 4 * gb, 8 * gb},
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"datasource", "datasource_type", "secure_socks_ds_proxy_enabled"},
+	)
+
+	datasourceRequestsInFlight = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "plugins",
+			Name:      "datasource_request_in_flight",
+			Help:      "A gauge of outgoing external data source requests currently being sent by Grafana",
+		},
+		[]string{"datasource", "datasource_type", "secure_socks_ds_proxy_enabled"},
+	)
+)
+
+// SanitizeLabelName removes all invalid chars from the label name.
+// If the label name is empty or contains only invalid chars, it
+// will return an error.
+func SanitizeLabelName(name string) (string, error) {
```

Comment:
Should not need to be exported, at least not here. Already lives here as well https://github.com/grafana/grafana-plugin-sdk-go/blob/d2fa15dd4660e763e91f06cc8cf26ca3e7b766ac/experimental/slo/metrics.go#L223